File: android/src/main/java/com/pspdfkit/react/NutrientViewRegistry.java
/*
 * Copyright © 2018-2026 PSPDFKit GmbH. All rights reserved.
 *
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE PSPDFKIT LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 */

package com.pspdfkit.react;

import android.util.Log;
import com.pspdfkit.views.PdfView;
import java.util.concurrent.ConcurrentHashMap;
import java.util.Map;

/**
 * Registry to track PdfView instances by their component reference ID.
 * This allows TurboModules to access specific view instances.
 */
public class NutrientViewRegistry {
    private static final String TAG = "NutrientViewRegistry";
    private static NutrientViewRegistry instance;
    private final Map<String, PdfView> views = new ConcurrentHashMap<>();
    
    private NutrientViewRegistry() {}
    
    public static synchronized NutrientViewRegistry getInstance() {
        if (instance == null) {
            instance = new NutrientViewRegistry();
        }
        return instance;
    }
    
    public void registerView(PdfView view, String nativeId) {
        if (view == null || nativeId == null) {
            Log.w(TAG, "Warning: Attempted to register view with null view or nativeId");
            return;
        }
        
        views.put(nativeId, view);
        Log.d(TAG, "Registered view with ID: " + nativeId + " (Total: " + views.size() + ")");
    }
    
    public PdfView getViewForId(String nativeId) {
        if (nativeId == null) {
            return null;
        }
        
        PdfView view = views.get(nativeId);
        if (view == null) {
            Log.w(TAG, "Warning: No view found for ID: " + nativeId);
        }
        
        return view;
    }
    
    public void unregisterView(String nativeId) {
        if (nativeId == null) {
            return;
        }
        
        PdfView removedView = views.remove(nativeId);
        if (removedView != null) {
            Log.d(TAG, "Unregistered view with ID: " + nativeId + " (Total: " + views.size() + ")");
        }
    }
    
    public int getRegisteredViewCount() {
        return views.size();
    }
}

File: ios/Common/NutrientViewRegistry.m
//
//  Copyright © 2018-2026 PSPDFKit GmbH. All rights reserved.
//
//  THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
//  AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE PSPDFKIT LICENSE AGREEMENT.
//  UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
//  This notice may not be removed from this file.
//

#import "NutrientViewRegistry.h"
#import "RCTPSPDFKitView.h"

@interface NutrientViewRegistry ()
@property (nonatomic, strong) NSMutableDictionary<NSString *, RCTPSPDFKitView *> *views;
@property (nonatomic, strong) dispatch_queue_t queue;
@end

@implementation NutrientViewRegistry

+ (instancetype)shared {
    static NutrientViewRegistry *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    if (self = [super init]) {
        _views = [NSMutableDictionary dictionary];
        _queue = dispatch_queue_create("com.pspdfkit.nutrientviewregistry", DISPATCH_QUEUE_SERIAL);
    }
    return self;
}

- (void)registerView:(RCTPSPDFKitView *)view withId:(NSString *)nativeId {
    if (!view || !nativeId) {
        NSLog(@"[NutrientViewRegistry] Warning: Attempted to register view with nil view or nativeId");
        return;
    }
    
    dispatch_async(self.queue, ^{
        self.views[nativeId] = view;
        NSLog(@"[NutrientViewRegistry] Registered view with ID: %@ (Total: %lu)", nativeId, (unsigned long)self.views.count);
    });
}

- (nullable RCTPSPDFKitView *)viewForId:(NSString *)nativeId {
    if (!nativeId) {
        return nil;
    }
    
    __block RCTPSPDFKitView *view = nil;
    dispatch_sync(self.queue, ^{
        view = self.views[nativeId];
    });
    
    if (!view) {
        NSLog(@"[NutrientViewRegistry] Warning: No view found for ID: %@", nativeId);
    }
    
    return view;
}

- (void)unregisterViewWithId:(NSString *)nativeId {
    if (!nativeId) {
        return;
    }
    
    dispatch_async(self.queue, ^{
        RCTPSPDFKitView *removedView = self.views[nativeId];
        if (removedView) {
            [self.views removeObjectForKey:nativeId];
            NSLog(@"[NutrientViewRegistry] Unregistered view with ID: %@ (Total: %lu)", nativeId, (unsigned long)self.views.count);
        }
    });
}

- (NSUInteger)registeredViewCount {
    __block NSUInteger count = 0;
    dispatch_sync(self.queue, ^{
        count = self.views.count;
    });
    return count;
}

@end 